std/zig/parse.zig

```diff
@@ -1,3659 +1,3044 @@
-                continue;
-            },
-            State.ParamDeclComma => |fn_proto| {
-                switch (expectCommaOrEnd(&tok_it, tree, Token.Id.RParen)) {
-                    ExpectCommaOrEndResult.end_token => |t| {
-                        if (t == null) {
-                            stack.append(State{ .ParamDecl = fn_proto }) catch unreachable;
-                        }
-                        continue;
-                    },
-                    ExpectCommaOrEndResult.parse_error => |e| {
-                        try tree.errors.push(e);
-                        return tree;
-                    },
-                }
-            },
+    // TODO: supply default value to struct field when ast.Node.StructField supports it
+    const default_value = blk: {
+        if (eatToken(it, .Equal)) |_| {
+            break :blk try expectNode(arena, it, tree, parseExpr, AstError{
+                .ExpectedExpr = AstError.ExpectedExpr{ .token = it.index },
+            });
+        } else break :blk null;
+    };
 
-            State.MaybeLabeledExpression => |ctx| {
-                if (eatToken(&tok_it, tree, Token.Id.Colon)) |_| {
-                    stack.append(State{
-                        .LabeledExpression = LabelCtx{
-                            .label = ctx.label,
-                            .opt_ctx = ctx.opt_ctx,
-                        },
-                    }) catch unreachable;
-                    continue;
-                }
+    switch (kind) {
+        .Keyword_struct => {
+            const node = try arena.create(Node.StructField);
+            node.* = Node.StructField{
+                .base = Node{ .id = .StructField },
+                .name_token = name_token,
+                .type_expr = type_expr orelse undefined,
+                .doc_comments = doc_comments,
+                .visib_token = null,
+            };
+            return &node.base;
+        },
+        .Keyword_union => {
+            const node = try arena.create(Node.UnionTag);
+            node.* = Node.UnionTag{
+                .base = Node{ .id = .UnionTag },
+                .doc_comments = doc_comments,
+                .name_token = name_token,
+                .type_expr = type_expr,
+                .value_expr = default_value,
+            };
+            return &node.base;
+        },
+        .Keyword_enum => {
+            const node = try arena.create(Node.EnumTag);
+            node.* = Node.EnumTag{
+                .base = Node{ .id = .EnumTag },
+                .doc_comments = doc_comments,
+                .name_token = name_token,
+                .value = default_value,
+            };
+            return &node.base;
+        },
+        else => unreachable,
+    }
+}
 
-                _ = try createToCtxLiteral(arena, ctx.opt_ctx, ast.Node.Identifier, ctx.label);
-                continue;
-            },
-            State.LabeledExpression => |ctx| {
-            
```

Comment:
I don't think we should encode in the grammar that only one `volatile` (or any other modifier) should be allowed on pointers. It would make the grammar more complicated than it has to be.

However, I think it should be an error. Grammars do not enforce semantics, and we can define that it is semantically illegal to have 2 `const` or whatever. Where we catch this semantic error is up to the implementation. I think the easiest place to catch it is probably in the parser in this function.

Reply:
Ok, I see what you mean. Originally I was thinking that the rule should be more like this:

    / ArrayTypeStart ByteAlign? KEYWORD_const? KEYWORD_volatile? KEYWORD_allowzero?

But that would enforce ordering, which would be a bit like prescribing an ordering to prefix operators. Which would be poopy.

I'll implement your suggestion of emitting an error. That seems especially sound given that multiple align qualifiers could conflict.